controllers/k8ssandra/secrets.go

```diff
@@ -4,28 +4,66 @@ import (
+	"github.com/k8ssandra/k8ssandra-operator/pkg/reaper"
 	"github.com/k8ssandra/k8ssandra-operator/pkg/result"
 	"github.com/k8ssandra/k8ssandra-operator/pkg/secret"
 	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
 )
 
 func (r *K8ssandraClusterReconciler) reconcileSuperuserSecret(ctx context.Context, kc *api.K8ssandraCluster, logger logr.Logger) result.ReconcileResult {
-	if kc.Spec.Cassandra.SuperuserSecretName == "" {
-		// Note that we do not persist this change because doing so would prevent us from
-		// differentiating between a default secret by the operator vs one provided by the
-		// client that happens to have the same name as the default name.
-		kc.Spec.Cassandra.SuperuserSecretName = secret.DefaultSuperuserSecretName(kc.Spec.Cassandra.Cluster)
-		logger.Info("Setting default superuser secret", "SuperuserSecretName", kc.Spec.Cassandra.SuperuserSecretName)
-	}
+	if kc.Spec.Auth {
+		if kc.Spec.Cassandra.Auth.SuperuserSecretRef.Name == "" {
+			// Note that we do not persist this change because doing so would prevent us from
+			// differentiating between a default secret by the operator vs one provided by the
+			// client that happens to have the same name as the default name.
+			kc.Spec.Cassandra.Auth.SuperuserSecretRef.Name = secret.DefaultSuperuserSecretName(kc.Spec.Cassandra.Cluster)
+			logger.Info("Setting default superuser secret", "SuperuserSecretName", kc.Spec.Cassandra.Auth.SuperuserSecretRef.Name)
+		}
 
-	if err := secret.ReconcileSecret(ctx, r.Client, kc.Spec.Cassandra.SuperuserSecretName, utils.GetKey(kc)); err != nil {
-		logger.Error(err, "Failed to verify existence of superuserSecret")
-		return result.Error(err)
+		if err := secret.ReconcileSecret(ctx, r.Client, kc.Spec.Cassandra.Auth.SuperuserSecretRef.Name, utils.GetKey(kc)); err != nil {
+			logger.Error(err, "Failed to verify existence of superuserSecret")
+			return result.Error(err)
+		}
 	}
 
 	return result.Continue()
 }
 
+func (r *K8ssandraClusterReconciler) reconcileReaperSecrets(ctx context.Context, kc *api.K8ssandraCluster, logger logr.Logger) result.ReconcileResult {
+	logger.Info("Reconciling Reaper user secrets")
+	if kc.Spec.Reaper != nil {
+		kcKey := utils.GetKey(kc)
+
+		// The CQL secret must only be reconciled when auth is enabled
+		if kc.Spec.Auth {
+			cassandraUserSecretRef := kc.Spec.Reaper.CassandraUserSecretRef
+			if cassandraUserSecretRef.Name == "" {
+				cassandraUserSecretRef.Name = reaper.DefaultUserSecretName(kc.Name)
+			}
+			logger = logger.WithValues("ReaperCassandraUserSecretRef", cassandraUserSecretRef)
+			if err := secret.ReconcileSecret(ctx, r.Client, cassandraUserSecretRef.Name, kcKey); err != nil {
+				logger.Error(err, "Failed to reconcile Reaper CQL user secret")
+				return result.Error(err)
+			}
+		} else {
+			logger.Info("Auth is disabled: skipping CQL user secret reconciliation")
+		}
+
+		// The JMX secret must be reconciled even when auth is disabled
```

Comment:
This was my guess but I am not sure here. Should we instead allow JMX remote connections without auth if auth is disabled globally?

Reply:
If Reaper is enabled, we need to enable remote JMX access. If we enable remote JMX access, I think it should always have JMX auth enabled as well regardless of whether Cassandra auth is enabled.

Reply:
This turned out to be way more complex than I imagined. 

The reason is that cassandra-env.sh recognizes the `LOCAL_JMX` env var, but local/remote toggling gets mixed with authentication toggling. Indeed to enable remote JMX access we need: `LOCAL_JMX=no`; but this also enables JMX auth :-(

In summary, using the `LOCAL_JMX` var:
- It is not possible to enable remote access WITHOUT auth;
- It is not possible to restrict access to localhost WITH auth.

The last bullet point is the most annoying IMO.

The only workaround that I could find is to use the `JVM_EXTRA_OPTS` env var. It would be probably nicer if we could expose `JVM_EXTRA_OPTS` in `CassandraDatacenter` CRD.

I tried to explain the whole thing here:

https://github.com/k8ssandra/k8ssandra-operator/blob/f221f13604589355cabe14c50e2d375aedc81470/pkg/cassandra/auth.go#L18-L30

Reply:
Would https://github.com/k8ssandra/k8ssandra/issues/323 help at all?

Reply:
Unfortunately no, however the changes in this PR will make it easier to switch to Cassandra internals for JMX authentication later on.

Note that IMO we'll need to use `JVM_EXTRA_OPTS` to do that as well.